File: 0231_Power_of_Two/Solution2.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    bool isPowerOfTwo(int n) {
        return n > 0 && (n & (n - 1)) == 0;
    }
};

void test(string test_name, int n, bool expected)
{
    Solution s;
    bool res = s.isPowerOfTwo(n);
    if (res == expected) {
        cout << test_name << " success." << endl;
    } else {
        cout << test_name << " failed." << endl;
    }
}

int main()
{
    int n1 = 1;
    bool expected1 = true;
    test("test1", n1, expected1);

    int n2 = 16;
    bool expected2 = true;
    test("test2", n2, expected2);

    int n3 = 218;
    bool expected3 = false;
    test("test3", n3, expected3);

    int n4 = 1073741825;
    bool expected4 = false;
    test("test4", n4, expected4);

    return 0;
}

File: of53_II_0～n-1中缺失的数字/Solution2.py
from typing import *

class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == mid:
                l += 1
            else:
                r -= 1
        return l


def test(test_name, nums, expected):
    res = Solution().missingNumber(nums)
    if res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')


if __name__ == "__main__":
    nums1 = [0,1,3]
    expected1 = 2
    test('test1', nums1, expected1)

    nums2 = [0,1,2,3,4,5,6,7,9]
    expected2 = 8
    test('test2', nums2, expected2)

    nums3 = [0]
    expected3 = 1
    test('test3', nums3, expected3)

    nums4 = [1]
    expected4 = 0
    test('test4', nums4, expected4)


# 一个长度为n-1的递增排序数组中的所有数字都是唯一的，并且每个数字都在范围0～n-1之内。
# 在范围0～n-1内的n个数字中有且只有一个数字不在该数组中，请找出这个数字。

# 示例 1:

# 输入: [0,1,3]
# 输出: 2

# 示例 2:
# 输入: [0,1,2,3,4,5,6,7,9]
# 输出: 8
#  

# 限制：
# 1 <= 数组长度 <= 10000
